File: hiob_contracts/factory/edge_registry.py
"""The mandatory semantic-edge registry — one row per directed Planet→Planet edge.

PRD_CREATIVE_FACTORY_HARMONY FR-2 and §5 (target factory flow). This is both the
Phase 0 "edge graph" artifact and the Phase 1 machine-readable authority that CI
checks: "CI rejects an unregistered direct edge" and "the runtime manifest and
live `.well-known` surfaces must match the registry."

Every row records the source output contract, the target input contract, the
Karma policy, the timeout, and the criticality. Binary uploads, control polling,
health probes, and human-approval commands are NOT semantic edges and are absent
here by design (§4.2).

Exact schema digests are resolved from each Planet's manifest at runtime; this
static registry binds *which* edges exist, *who* consumes, and *under which
policy* — the topology CI enforces.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

Criticality = Literal["required", "optional"]


@dataclass(frozen=True)
class SemanticEdge:
    """One directed, Karma-mediated handoff between two Planets."""

    edge_id: str
    source_planet: str
    source_contract: str
    target_planet: str
    target_node_id: str
    target_contract: str
    policy_id: str
    timeout_ms: int
    criticality: Criticality
    note: str = ""


# ── The factory edge graph (§5). Order follows the flow. ─────────────────────
EDGES: tuple[SemanticEdge, ...] = (
    # Planning slice (Phase 2): Janus → Parzifal → Ares.
    SemanticEdge(
        "j2p", "janus", "JanusBrief", "parzifal", "parzifal.target.consolidate",
        "ParzifalTargetInput", "policy.j2p", 30_000, "required",
        "Janus evidence brief refined for Parzifal target/identity derivation",
    ),
    SemanticEdge(
        "p2a", "parzifal", "TargetProfile+IdentityLock+CastSheet", "ares", "ares.script.build",
        "AresScriptInput", "policy.p2a", 30_000, "required",
        "Parzifal identity/target refined for Ares; Ares cannot run without this receipt (FR-3)",
    ),
    # Media plan slice (Phase 3): Ares → {Athena, Orpheus, Apollo} plans.
    SemanticEdge(
        "a2athena", "ares", "ScriptPackage+BeatPlan", "athena", "athena.plan",
        "AthenaPlanInput", "policy.a2media", 30_000, "required",
        "Ares beats refined for Athena visual planning",
    ),
    SemanticEdge(
        "a2orpheus", "ares", "ScriptPackage+BeatPlan", "orpheus", "orpheus.plan",
        "OrpheusPlanInput", "policy.a2media", 30_000, "required",
        "Ares beats refined for Orpheus voice/music planning (voice required)",
    ),
    SemanticEdge(
        "a2apollo", "ares", "ScriptPackage+BeatPlan", "apollo", "apollo.plan",
        "ApolloPlanInput", "policy.a2media", 30_000, "optional",
        "Ares beats refined for Apollo SFX planning (optional enhancement, §9.1)",
    ),
    # Fan-in (Phase 3/4): materialized media bundle → Atropos draft.
    SemanticEdge(
        "media2atropos", "athena+orpheus+apollo",
        "VisualArtifactSet+VoiceMusicArtifactSet+SfxArtifactSet",
        "atropos", "atropos.draft", "AtroposDraftInput", "policy.media2atropos",
        45_000, "required",
        "Required-media fan-in refined into an Atropos draft snapshot input",
    ),
    # Editorial loop (Phase 4) — SUNSET 2026-07-21.
    # Artemis identity is product/evidence seal only (artemis.references.snapshot).
    # Edges kept for contract validation / historical receipts; criticality optional
    # so they never block F1. Hard-delete after consumer cleanup (see hiob-artemis/docs/SUNSET.md S7).
    SemanticEdge(
        "atropos2artemis", "atropos", "CompositionSnapshot", "artemis", "artemis.review",
        "ArtemisReviewInput", "policy.atropos2artemis", 45_000, "optional",
        "SUNSET: not on F1 spine; Artemis = product/evidence seal only (was editorial QA)",
    ),
    SemanticEdge(
        "artemis2atropos", "artemis", "EditorialProposal", "atropos", "atropos.apply",
        "AtroposApplyInput", "policy.artemis2atropos", 30_000, "optional",
        "SUNSET: human-accepted editorial proposals (optional; not live produce path)",
    ),
    # Final render (Phase 5): carries the G3 ApprovalReceipt reference, not a boolean.
    SemanticEdge(
        "atropos2hephaestus", "atropos", "CompositionSnapshot", "hephaestus", "hephaestus.render",
        "HephaestusRenderInput", "policy.atropos2hephaestus", 60_000, "required",
        "Final snapshot + G3 approval reference refined for render authorization (FR-8)",
    ),
)

_BY_ID: dict[str, SemanticEdge] = {e.edge_id: e for e in EDGES}


def get_edge(edge_id: str) -> SemanticEdge | None:
    """Look up a registered edge by id, or None."""
    return _BY_ID.get(edge_id)


def is_registered_edge(source_planet: str, target_planet: str) -> bool:
    """True iff a semantic edge from `source_planet` to `target_planet` is registered.

    CI uses this to reject any direct Planet→Planet consumption that is not a
    declared, Karma-mediated edge (FR-2). Fan-in sources are recorded as a joined
    `a+b+c` source_planet token; membership is checked per component.
    """
    for e in EDGES:
        sources = e.source_planet.split("+")
        if target_planet == e.target_planet and source_planet in sources:
            return True
    return False


def required_edges() -> tuple[SemanticEdge, ...]:
    """Edges whose failure must block the run (fan-in cannot advance without them)."""
    return tuple(e for e in EDGES if e.criticality == "required")

File: ts/src/factory/edge-registry.ts
/**
 * Semantic-edge registry — TS mirror of factory/edge_registry.py.
 *
 * ⚠️ AUTHORITY: Python. 엣지 목록·criticality는 Python `EDGES`와 일치해야 한다.
 * FR-2: CI가 미등록 직결 엣지를 거부하고, 런타임 manifest/`.well-known`이 이 표와 일치.
 */

export type Criticality = 'required' | 'optional';

export interface SemanticEdge {
  edge_id: string;
  source_planet: string;
  source_contract: string;
  target_planet: string;
  target_node_id: string;
  target_contract: string;
  policy_id: string;
  timeout_ms: number;
  criticality: Criticality;
  note?: string;
}

export const EDGES: readonly SemanticEdge[] = [
  { edge_id: 'j2p', source_planet: 'janus', source_contract: 'JanusBrief',
    target_planet: 'parzifal', target_node_id: 'parzifal.target.consolidate',
    target_contract: 'ParzifalTargetInput', policy_id: 'policy.j2p', timeout_ms: 30000,
    criticality: 'required' },
  { edge_id: 'p2a', source_planet: 'parzifal', source_contract: 'TargetProfile+IdentityLock+CastSheet',
    target_planet: 'ares', target_node_id: 'ares.script.build',
    target_contract: 'AresScriptInput', policy_id: 'policy.p2a', timeout_ms: 30000,
    criticality: 'required' },
  { edge_id: 'a2athena', source_planet: 'ares', source_contract: 'ScriptPackage+BeatPlan',
    target_planet: 'athena', target_node_id: 'athena.plan',
    target_contract: 'AthenaPlanInput', policy_id: 'policy.a2media', timeout_ms: 30000,
    criticality: 'required' },
  { edge_id: 'a2orpheus', source_planet: 'ares', source_contract: 'ScriptPackage+BeatPlan',
    target_planet: 'orpheus', target_node_id: 'orpheus.plan',
    target_contract: 'OrpheusPlanInput', policy_id: 'policy.a2media', timeout_ms: 30000,
    criticality: 'required' },
  { edge_id: 'a2apollo', source_planet: 'ares', source_contract: 'ScriptPackage+BeatPlan',
    target_planet: 'apollo', target_node_id: 'apollo.plan',
    target_contract: 'ApolloPlanInput', policy_id: 'policy.a2media', timeout_ms: 30000,
    criticality: 'optional' },
  { edge_id: 'media2atropos', source_planet: 'athena+orpheus+apollo',
    source_contract: 'VisualArtifactSet+VoiceMusicArtifactSet+SfxArtifactSet',
    target_planet: 'atropos', target_node_id: 'atropos.draft',
    target_contract: 'AtroposDraftInput', policy_id: 'policy.media2atropos', timeout_ms: 45000,
    criticality: 'required' },
  // SUNSET 2026-07-22: artemis.review HTTP removed (quarantined). Keep optional for registry parity with Python; do not re-require on F1.
  { edge_id: 'atropos2artemis', source_planet: 'atropos', source_contract: 'CompositionSnapshot',
    target_planet: 'artemis', target_node_id: 'artemis.review',
    target_contract: 'ArtemisReviewInput', policy_id: 'policy.atropos2artemis', timeout_ms: 45000,
    criticality: 'optional', note: 'SUNSET — artemis.review 404; not on F1 spine' },
  { edge_id: 'artemis2atropos', source_planet: 'artemis', source_contract: 'EditorialProposal',
    target_planet: 'atropos', target_node_id: 'atropos.apply',
    target_contract: 'AtroposApplyInput', policy_id: 'policy.artemis2atropos', timeout_ms: 30000,
    criticality: 'optional', note: 'SUNSET — not on F1 produce spine' },
  { edge_id: 'atropos2hephaestus', source_planet: 'atropos', source_contract: 'CompositionSnapshot',
    target_planet: 'hephaestus', target_node_id: 'hephaestus.render',
    target_contract: 'HephaestusRenderInput', policy_id: 'policy.atropos2hephaestus', timeout_ms: 60000,
    criticality: 'required' },
];

const BY_ID = new Map(EDGES.map((e) => [e.edge_id, e]));

export function getEdge(edgeId: string): SemanticEdge | undefined {
  return BY_ID.get(edgeId);
}

/** source→target 방향 시맨틱 엣지가 등록됐으면 true (fan-in은 `a+b+c` 성분별 검사). */
export function isRegisteredEdge(sourcePlanet: string, targetPlanet: string): boolean {
  return EDGES.some(
    (e) => e.target_planet === targetPlanet && e.source_planet.split('+').includes(sourcePlanet)
  );
}

export function requiredEdges(): SemanticEdge[] {
  return EDGES.filter((e) => e.criticality === 'required');
}
